File: 0130-surrounded-regions/solution.java
class Pair{
    int first, second;
    public Pair(int f, int s){
        first = f;
        second = s;
    }
}
class Solution {
    public void solve(char[][] arr) {
        int n = arr.length;
        int m = arr[0].length;
        int [][]visited = new int[n][m];
        for( int i=0; i<n; i++){
            for( int j=0; j<m; j++){
                if( i==0 || j==0 || i == n-1 || j == m-1){
                    if( arr[i][j] == 'O'  &&  visited[i][j] == 0)
                         markVisited(i, j, visited, arr, n, m);
                }
            }
        }

        for( int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if( arr[i][j] == 'O'  && visited[i][j] == 0){
                    arr[i][j] = 'X';
                }
            }
        }
    }
    public void markVisited(int i, int j, int [][]visited, char[][] arr, int n,             int m){

        visited[i][j] = 1;
        //queue<pair<int, int>> q;
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(i, j));


        int drow[] = { 0, +1, 0,-1};
        int dcol[] = {+1, 0, -1, 0};
        while( !q.isEmpty()){
            Pair p = q.remove();
            int row = p.first;
            int col = p.second;
            //q.remove();

            for( int k=0; k<4; k++){
                int nrow = row + drow[k];
                int ncol = col + dcol[k];

                if( nrow >=0 && nrow<n && ncol >=0 && ncol <m && 
                           visited[nrow][ncol] == 0  && arr[nrow][ncol] == 'O' ){

                     visited[nrow][ncol] = 1;
                     q.add(new Pair(nrow,ncol));
                }
            }
        }
    }
}

File: 0130-surrounded-regions/solution.cpp
class Solution {
    void dfs(int row, int col, vector<vector<char>> &board, vector<vector<int>> &visited, int n, int m){
        int drow[] = {-1, 0, 1, 0};
        int dcol[] = {0, 1, 0, -1};
        
        for(int i=0; i<=3; i++){
            int nrow = drow[i] + row;
            int ncol = dcol[i] + col;
            if(nrow <n-1 && nrow >0 && ncol <m-1 && ncol >0 && board[nrow][ncol] == 'O' && visited[nrow][ncol] == 0){
                visited[nrow][ncol] = 1;
                dfs(nrow, ncol, board, visited, n, m);
            }
        }
    }
public:
    void solve(vector<vector<char>>& board) {
        int n = board.size();
        int m = board[0].size();
        vector<vector<int>> visited(n, vector<int>(m, 0));

        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if((i == 0 || i == n-1 || j == 0 || j == m-1) && board[i][j] == 'O'){
                    // cout<<i<<', '<<j<<'    ';
                    visited[i][j] = 1;
                    dfs(i, j, board, visited, n, m);
                }
            }
        }

        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(board[i][j] == 'O' && visited[i][j] == false) 
                    board[i][j] = 'X';
            }
        }
    }
};
